lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,247 @@
+"""
+Models used for robust grading.
+
+Robust grading allows student scores to be saved per-subsection independent
+of any changes that may occur to the course after the score is achieved.
+"""
+
+from django.db import models
+from django.db.utils import IntegrityError
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import CourseKeyField, UsageKeyField
+
+from base64 import b64encode
+from collections import namedtuple
+from hashlib import sha256
+import json
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+# Used to serialize information about a block at the time it was used in
+# grade calculation.
+BlockRecord = namedtuple('BlockRecord', ['locator', 'weight', 'max_score'])
+
+
+def blockrecord_list_to_json(blocks):
+    """
+    Return a JSON-serialized version of the list of block records.
+    """
+    return json.dumps([block._asdict() for block in blocks], separators=(',', ':'), sort_keys=True)
```

Comment:
My main thought in suggesting this was more to be completely explicit about what the JSON formatting is supposed to be, since we're relying on the JSON output for reproducible hashes. Choosing those values specifically was to make it as small as possible, since it doesn't really make a difference anyway.